### ISSUE TITLE
Zip2john problem after 8246d20

### ISSUE THREAD

**Issue by ghost** [openwall/john#4025]
I do have libz. I have it working fine with release 0dd0fd3. The problem is bleeding edge and everything after 8246d20. 
I'm unsure if I can share the file (would have to check), but regardless, it's 3.5GB, so that would likely be a problem regardless.

I spent a little more time looking at the 8246d20 commit. If you look, it moves the code with the comment "// If bit 3 of flags is set" into the function "scan_for_eod". But scan_for_eod is only called if both cmp_len && decomp_len are 0 (line 487).
Turns out, in my case, cmp_len is 0 but decomp_len isn't. And therefore, scan_for_eod isn't called and therefore cmp_len is never set. 

I confirmed this by changing line 487 to if(cmp_len == 0 || decomp_len == 0). And now everything works. 

In other words, before 8246d20, scan_for_eod is always called, but only does things if data descriptor  0x08074b50UL or next local header 0x04034b50UL. 
After 8246d20, scan_for_eod is only called if cmp_len and decomp_len are 0 and still searches for those 2 headers. 

Seems like the if at line 486 could be removed. Or maybe the && replaced with ||. Not sure.

Hope this helps.

_Originally posted by @jwallior in https://github.com/magnumripper/JohnTheRipper/issues/2193#issuecomment-500081871_

**Comment by jwallior** [comment]
One of those 2 patches work for me. I'm unsure what's the right thing to do.

```
diff --git a/src/zip2john.c b/src/zip2john.c
index 88456f1d2..d8e1fadf9 100644
--- a/src/zip2john.c
+++ b/src/zip2john.c
@@ -663,7 +663,7 @@ static int LoadZipBlob(FILE *fp, zip_ptr *p, zip_file *zfp, const char *zip_fnam
                else if (extrafield_length)
                        fseek(fp, extrafield_length, SEEK_CUR);

-               if (p->cmp_len == 0 && p->decomp_len == 0)
+               if (p->cmp_len == 0 || p->decomp_len == 0)
                        scan_for_eod(&fp, p, size64);

                if (only_fname && strcmp(only_fname, (char*)filename)) {
```

```
diff --git a/src/zip2john.c b/src/zip2john.c
index 88456f1d2..1f680855d 100644
--- a/src/zip2john.c
+++ b/src/zip2john.c
@@ -663,8 +663,7 @@ static int LoadZipBlob(FILE *fp, zip_ptr *p, zip_file *zfp, const char *zip_fnam
                else if (extrafield_length)
                        fseek(fp, extrafield_length, SEEK_CUR);

-               if (p->cmp_len == 0 && p->decomp_len == 0)
-                       scan_for_eod(&fp, p, size64);
+               scan_for_eod(&fp, p, size64);

                if (only_fname && strcmp(only_fname, (char*)filename)) {
                        fseek(fp, p->cmp_len, SEEK_CUR);
```

**Comment by magnumripper** [comment]
@jwallior thanks a lot for reporting, this is excellent input. I will try to fix the problem by dry running, within a few days.

**Comment by magnumripper** [comment]
> ```diff
> diff --git a/src/zip2john.c b/src/zip2john.c
> index 88456f1d2..d8e1fadf9 100644
> --- a/src/zip2john.c
> +++ b/src/zip2john.c
> @@ -663,7 +663,7 @@ static int LoadZipBlob(FILE *fp, zip_ptr *p, zip_file *zfp, const char *zip_fnam
>                 else if (extrafield_length)
>                         fseek(fp, extrafield_length, SEEK_CUR);
> 
> -               if (p->cmp_len == 0 && p->decomp_len == 0)
> +               if (p->cmp_len == 0 || p->decomp_len == 0)
>                         scan_for_eod(&fp, p, size64);
> 
>                 if (only_fname && strcmp(only_fname, (char*)filename)) {
> ```

From a first look, this is the correct fix.

**Comment by magnumripper** [comment]
@jwallior could you please post the full stderr output from running zip2john on the failing file, first *with* the bug (latest bleeding) and then *without* the bug (eg. changing `&&` to `||` per above). You can obfuscate file names if needed, I'm just interested in the output of flags/headers/sizes.

**Comment by jwallior** [comment]
Here is the data. I did change the filenames. One thing that I hadn't noticed before, with the patch zip2john finds 707 files in the zip and without the patch, it only finds 112. 

I only put the full line for the union of the files in the digest. decmplen seems to be correct even without the fix. (but cmplen and crc are wrong).

Hope this helps. If you need to full files (707 of them), let me know and I can add them. But it's a bit painful to get the names out and I thought that might be enough.

```
bleeding stdout
filename.zip:$pkzip2$3*2*1*0*8*0*33c2*33c2**1*0*8*0*4bb3*4bb3**2*0*0*12fda*6abd0000*0*4d*8*0*6abd*6abd**$/pkzip2$::filename.zip:f1, f2, f3:/hostname/filename.zip
bleeding stderr
ver 2.0 efh 5455 efh 7875 filename.zip/f1 PKZIP Encr: 2b chk, TS_chk, cmplen=0, decmplen=77786, crc=6ABD0000
ver 2.0 efh 5455 efh 7875 filename.zip/f2 PKZIP Encr: 2b chk, TS_chk, cmplen=0, decmplen=3124, crc=33C20000
ver 2.0 efh 5455 efh 7875 filename.zip/f3 PKZIP Encr: 2b chk, TS_chk, cmplen=0, decmplen=2819, crc=4BB30000
...
NOTE: It is assumed that all files in each archive have the same password.
If that is not the case, the hash may be uncrackable. To avoid this, use
option -o to pick a file at a time.

fix (&& -> ||) stdout
filename.zip:$pkzip2$3*2*1*0*8*24*13ab*2ba8*e11df20c793d56f172fb5155d7e72e6ad4fec3f06494f75923d58ae227b0d3fa1a9d440e*1*0*8*24*cd9e*6abd*937e42db005e6fd8c801a166417b3ce43dc7bda5e504042d103c072a5824e8f4d71c36a5*2*0*17*14*5f0c16fa*e7eae*51*8*17*5f0c*3bbb*b79e5b1fbbfc1f724cb926e460f0580cddc7d05799ae86*$/pkzip2$::filename.zip:f4, f5, f1:/hostname/filename.zip
fix (&& -> ||) stderr
ver 2.0 efh 5455 efh 7875 Scanning for EOD... FOUND Extended local header
filename.zip/f1 PKZIP Encr: 2b chk, TS_chk, cmplen=3218, decmplen=77786, crc=CD9EAA6E
ver 2.0 efh 5455 efh 7875 Scanning for EOD... FOUND Extended local header
filename.zip/f5 PKZIP Encr: 2b chk, TS_chk, cmplen=2852, decmplen=2974, crc=13AB8538
ver 2.0 efh 5455 efh 7875 Scanning for EOD... FOUND Extended local header
filename.zip/f2 PKZIP Encr: 2b chk, TS_chk, cmplen=3028, decmplen=3124, crc=6E33362C
ver 2.0 efh 5455 efh 7875 Scanning for EOD... FOUND Extended local header
filename.zip/f3 PKZIP Encr: 2b chk, TS_chk, cmplen=2685, decmplen=2819, crc=9D8E7AF
ver 2.0 efh 5455 efh 7875 Scanning for EOD... FOUND Extended local header
filename.zip/f4 PKZIP Encr: 2b chk, TS_chk, cmplen=23, decmplen=20, crc=5F0C16FA
...
NOTE: It is assumed that all files in each archive have the same password.
If that is not the case, the hash may be uncrackable. To avoid this, use
option -o to pick a file at a time.
```

**Comment by magnumripper** [comment]
Thanks. I pushed a similar fix. And thanks again for excellent input, it saved us quite some work figuring things out.
